### PR TITLE
[FLINK-29665][runtime] Support flexible subpartion range division

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change
Support flexible subpartion range division

## Verifying this change
Add some tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoStore.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        checkNotNull(info);
+
+        vertexInputInfos.compute(
+                jobVertexId,
+                (ignored, inputInfos) -> {
+                    if (inputInfos == null) {
+                        inputInfos = new HashMap<>();
+                    }
+
+                    // Note that a job vertex can have 2 inputs with the same IntermediateDataSetID
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoStore.java:54]: Here is mainly to explain why we use `putIfAbsent`. I will add the "if a vertex has multiple job edges connecting to the same result, their distribution pattern must be the same and therefore the VertexInputInfo will be the same" to the doc of VertexInputInfoStore.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+        @Override
+        public int getNumPartitions() {
+            return intermediateResult.getNumberOfAssignedPartitions();
+        }
+
+        @Override
+        public int getNumSubpartitions(int partitionIndex) {
+            boolean isDynamicGraph = intermediateResult.getProducer().getGraph().isDynamic();
+            // Note that for non-dynamic graph, the num of subpartition has not been decided at
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java:251]: This comment should indeed be `non-dynamic`. Currently for non-dynamic graph, the number of subpartitions is decided by the connection information in the `EdgeManager`. At this point, this vertex has not been connected to predecessors(ExecutionJobVertex#connectToPredecessors has not be called), so the edge manager does not have the connection information, so we use UNKONWN here.

**Comment by wanglijie95** [comment]
Thanks for review @zhuzhurk. I 've addressed all comments, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java:259]: If we never expect a non-dynamic graph to invoke this method, it's better to add a check here, also some comments to explain it.

[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java]: Fixed

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java [hunk 3]
```diff
@@ -255,12 +259,11 @@ public int getNumPartitions() {
 
         @Override
         public int getNumSubpartitions(int partitionIndex) {
-            boolean isDynamicGraph = intermediateResult.getProducer().getGraph().isDynamic();
-            // Note that for non-dynamic graph, the num of subpartition has not been decided at
-            // this time
-            return isDynamicGraph
-                    ? intermediateResult.getPartitions()[partitionIndex].getNumberOfSubpartitions()
-                    : IntermediateResultPartition.NUM_SUBPARTITIONS_UNKNOWN;
+            // Note that this method should only be called for dynamic graph
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java:262]: Better to explain why, e.g. this method is used to compute which sub-partitions a consumer vertex should consume, however, for non-dynamic graph it is not needed, and the number of sub-partitions is not decided at this stage, due to the execution edge are not created.

Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/VertexInputInfoComputationUtils.java]: Added

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request closed without merging**